File: 2020-21/contest2/prog/intermediate/lex_strings.py
import os
import sys

while(True): # for batch process - do not do this in your hackerrank.com solutions
	s = ''
	m = 0
	args = []
	try:
		args = [x for x in input().split()]
	except Exception as e:
		break
	s = args[0]
	m = int(args[1])
	block = [] # Python strings are immutable - use lists as string builders
	# key: string length, val: strings with the same length - in descending order
	store = {}
	for c in s:
		if len(block) > 0 and block[len(block)-1] != c:
			if not store.get(len(block)):
				store[len(block)] = []
			store[len(block)].append(block);
			block = []
		block.append(c);
	# don't forget about the leftover
	if not store.get(len(block)):
		store[len(block)] = []
	store[len(block)].append(block);
	# combine the blocks
	add = []
	for k in reversed(sorted(store.keys())):
		strs = sorted(store.get(k))
		for s in strs:
			for c in s:
				add.append(c)
	# compress the chained
	length = 0
	res = []
	for cur in add: # compress
		if len(res) > 0 and cur == res[len(res)-1]:
			length += 1
			if length <= m:
				res.append(cur)
		else:
			length = 1
			res.append(cur)
	print(''.join(res))

File: 2020-21/contest2/prog/intermediate/lex_strings.java
import java.io.*;
import java.util.*;

public class lex_strings {
	public static void main(String[] args) {
		Scanner input = new Scanner(System.in);
		while(true) { // for batch process - do not do this in your hackerrank.com solutions
			String s; int m;
			try {
				s = input.next();
				m = input.nextInt();
			} catch(Exception e) {
				break; // exit the program
			}
			// slice string s into blocks and save to a descendant storage
			StringBuilder block = new StringBuilder();
			// key: string length, val: strings with the same length - in descending order
			TreeMap<Integer,ArrayList<String>> store =
				new TreeMap<Integer,ArrayList<String>>(Collections.reverseOrder());
			for(int i = 0; i < s.length(); ++i) {
				char c = s.charAt(i);
				if(block.length()>0 && block.charAt(block.length()-1) != c) {
					ArrayList<String> arr;
					if(store.get(block.length())==null) {
						arr = new ArrayList<String>();
						store.put(block.length(), arr);
					} else {
						arr = store.get(block.length());
					}
					arr.add(block.toString());
					block = new StringBuilder();
				}
				block.append(c);
			}
			// don't forget about the leftover
			ArrayList<String> arr;
			if(store.get(block.length())==null) {
				arr = new ArrayList<String>();
				store.put(block.length(), arr);
			} else {
				arr = store.get(block.length());
			}
			arr.add(block.toString()); 
			// combine the blocks
			StringBuilder add = new StringBuilder();
			for(Map.Entry<Integer,ArrayList<String>> entry : store.entrySet()) {
				ArrayList<String> strs = entry.getValue();
				Collections.sort(strs);
				for(int i = 0; i < strs.size(); ++i) {
					add.append(strs.get(i));
				}
			}
			// compress the chained
			int len = 0;
			StringBuilder res = new StringBuilder();
			for(int i = 0; i < add.length(); ++i) { // compress
				char cur = add.charAt(i);
				if(res.length() != 0 && cur == res.charAt(res.length()-1)) {
					++len;
					if(len <= m)
						res.append(cur);
				} else {
					len = 1;
					res.append(cur);
				}
			}
			System.out.println(res.toString());
		}
	}
}
